### ISSUE TITLE
Traffic Signal System: observer design pattern improper usage

### ISSUE THREAD

**Issue by siddharthmagadum16** [ashishps1/awesome-low-level-design#62]
Hi Ashish, I was going through LLD of Traffic Signal System. The design uses observer pattern, but I think that this is slightly misimplemented.
<br>
The Code [here](https://github.com/ashishps1/awesome-low-level-design/blob/995b0fb4ece313a8147d6a06aea2101bb16017e7/solutions/java/src/trafficsignalsystem/TrafficLight.java#L20), i.e. `notifyObservers()` should not be present in TrafficLight entity. <br>
Rather, it should be called in TrafficController entity, as it has the list of Roads (indirectly TrafficLights) and acts as a subject to infrom all TrafficLights about the signal color update of each of them.

<br>
In short, `notifyObservers()`  method should be a part of TrafficController and `listenTrafficEvent()/changeColor()`  should be part of TrafficLight.

**Comment by ashishps1** [comment]
changed the implementation to use state design pattern so this issue is no longer there.
